keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,699 @@
+In 3329, there was a decision to make kubelet transition pods to failed before deleting them.  This is feature toggled guarded by `PodDisruptionCondition`.  
+This means that when this feature is turned on, the job controller can count pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  So this pod would be considered active until it is fully terminated.  
+If `PodDisruptionCondition` is turned off then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.
+This causes some problems in tracking for active versus failed because we have different behavior based on `PodDisruptionCondition`.  To migitate this, we decided to add a new field called `terminating`.  This will track terminating pods separately from active and failed.  
+
+Another issue is described here: https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509.  
+If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.  Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.  We will need to modify the PodGC (gc_controller.go) in the case of `JobRecreatePodsFailed` being enabled while `PodDisruptionConditions` is disabled.  
+
+## Design Details
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new enum field:
+
+```golang
+// This field controls when we recreate pods
+// Default will be TerminatingOrFailed ie recreate pods when they are failed
+// +enum 
+type RecreatePodsWhen string
+const (
+ // This is a field that recreates pods when they are marked as terminating or failed
+ // ie this means that as soon as pods get marked for deletion they will be recreated
+ TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
+ // Only recreate pods when they are marked as failed.
+ Failed              RecreatePodsWhen = "Failed"
+)
+```
+
+```golang
+type JobSpec struct{
+  ...
+ // RecreatePodsWhen specifies when pods should be recreated.
+ // TerminatingOrFailed means to recreate when a pod is either terminating or failed
+ // TerminatingOrFailed is the default.
+ // Failed means to wait until pods are fully terminated or failed before recreating
+ // +optional
+ RecreatePodsWhen *RecreatePodsWhen
+}
+```
+
+So we can count terminating pods separately from active or failed we need to include a new field in the JobStatus.
+
+```golang
+type JobStatus struct {
+  ...
+  // Number of terminating pods
+  // +optional
+  terminating *int32
```

Comment:
That's still debatable, imo.

Reply:
I hope my updates clear this up.

Reply:
It can. But that's extra memory to keep the watch for Pods.

Ideally, Kueue can take decisions based entirely on the Job object.

Reply:
OTOH, I think a `terminating` field is valuable for users. They might be wondering: why is `active` less than `parallelism`? And the answer is: some pods are terminating, which is not currently visible.